include/dr/mhp/containers/distributed_dense_matrix.hpp

```diff
@@ -0,0 +1,496 @@
+        index < (local_rows_ind_.second + 1) * (int)shape_[1]) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+private:
+  friend d_segment_iterator<distributed_dense_matrix>;
+  friend dm_rows<distributed_dense_matrix>;
+  friend dm_rows_iterator<distributed_dense_matrix>;
+  friend subrange_iterator<distributed_dense_matrix>;
+
+  key_type shape_; // matrix shape
+  key_type segment_shape_;
+
+  std::size_t segment_size_ = 0; // size of local data
+  std::size_t data_size_ = 0;    // all data - local + halo
+
+  T *data_ = nullptr; // data ptr
+
+  span_halo<T> *halo_ = nullptr;
+  // halo boundaries counted in cells and rows
+  halo_bounds halo_bounds_, halo_bounds_rows_;
+
+  std::vector<d_segment<distributed_dense_matrix>> segments_;
+
+  // vector of "regular" rows in segment
+  dm_rows<distributed_dense_matrix> dm_rows_;
+  // rows in halo areas, prev & next
+  dm_rows<distributed_dense_matrix> dm_halo_p_rows_, dm_halo_n_rows_;
+
+  // global indices of locally stored rows (lowest and highest)
+  std::pair<difference_type, difference_type> local_rows_ind_ =
+      std::pair(-1, -1);
+
+  dr::rma_window win_;
+  Allocator allocator_;
+}; // class distributed_dense_matrix
+
+//
+// for_each, but iterator is passed to lambda
+// FIXME: enable sycl
+//
+template <typename T>
+auto for_each(dm_rows<distributed_dense_matrix<T>> &rows, auto op) {
+  for (auto itr = rng::begin(rows); itr != rng::end(rows); itr++) {
+    if ((*itr).segment()->is_local()) {
+      op(*itr);
+    }
+  }
+};
+
+#if 0 // temporary disabled, to be fixed
+template <typename T>
+concept has_segment = requires(T t) { (*(t.begin())).segment(); };
+
+template <typename T1, typename T2>
+  requires(has_segment<T1> && has_segment<T2>)
+auto for_each(zip_view<T1, T2> &v, auto op) {
+  for (auto itr = rng::begin(v); itr != rng::end(v); itr++) {
+    auto [in, out] = *itr;
+    if (in.segment()->is_local()) {
+      assert(out.segment()->is_local());
+      op(*itr);
+    }
+  }
+};
+#endif
+//
+// transform, iterator pointing at element is passed to lambda
+//
+template <typename DM>
+void transform(dr::mhp::subrange<DM> &in, subrange_iterator<DM> out, auto op) {
+  for (subrange_iterator<DM> i = rng::begin(in); i != rng::end(in); i++) {
+    if (i.is_local()) {
+      assert(out.is_local());
+      *(out) = op(i);
+    }
+    ++out;
+  }
+}
+
+template <typename DM>
+void transform(rng::subrange<dm_rows_iterator<DM>> &in,
+               dm_rows_iterator<DM> out, auto op) {
+  for (auto i = rng::begin(in); i != rng::end(in); i++) {
+    if (i.is_local()) {
+      assert(out.is_local());
+      *out = op(i);
+    }
+    ++out;
+  }
+}
+
+} // namespace dr::mhp
+
+// Needed to satisfy rng::viewable_range
+
+template <typename T>
+inline constexpr bool rng::enable_borrowed_range<
+    dr::mhp::dm_rows<dr::mhp::distributed_dense_matrix<T>>> = true;
+
+template <typename T>
+inline constexpr bool rng::enable_borrowed_range<
```

Comment:
You won't need this if subrange inherits from view_interface

Reply:
Subrange modified, so I removed the commented code